### ISSUE TITLE
easier install needed

### ISSUE THREAD

**Issue by simonmichael** [alphapapa/matrix-client.el#91]
This sounds great, I'd like to try it out, but my installation's not going too well:

- Not found in MELPA. So I git clone.

- I just fixed up my emacs installation and I don't want to risk running some non-standard install script that installs another emacs

- I've never heard of quelpa, so I check it out and their README says it's retired - sounds like a dead end

- So it's manual installation for me. I try my usual fumble-around-with-dired technique: open the directory, select all .el files, B, L. 6 files fail to compile. The log shows lots of warnings and errors but I can't see which packages it might be needing.

- I read and follow the manual instructions more carefully. I add the matrix-client.el directory to load-path and try (require 'matrix-client). Now I see it's failing on require(ov). I work through these errors / look at the source and install a bunch of things with list-packages: a, anaphora, dash-functional, esxml, ov, request tracking. 

- Now it's getting further but fails thusly:
```
Debugger entered--Lisp error: (void-variable notify)
  byte-code("\304\010\305\306\307\310\011\n\311\312$\211\203\033\0\313\013\002\"\210\314\315\n\"\202_\0\316\013!\211\317\001\320\"\262\001\211\013\001H\262\001\211\002\004\316\003!\211\317\001\321\"\262\001\211\005\001H\262\001\211\002\004\002\322\001\011\323\312$@\324\325\326\327\011\"\"\314\330\003\003#\266\203\266\203\266\203\266\203\266\203\262\001%\207" [notify matrix-client-notification-rules input room matrix-client-def-room-command :docstring "Set room notification setting.\nWithout argument, displays help and current setting." :insert map-elt nil equal matrix-client-set-notification-rule format "Notification rule set: %s." eieio-pcase-slot-index-table eieio-pcase-slot-index-from-index-table client-data notification-rules cl-rassoc :test s-join ", " -map car "Current notification rule: %s.  Available rules: %s."] 25)
  require(matrix-notifications)
  eval-buffer(#<buffer  *load*> nil "/Users/simon/src/CHAT/matrix-client.el/matrix-client.el" nil t)  ; Reading at buffer position 1889
  load-with-code-conversion("/Users/simon/src/CHAT/matrix-client.el/matrix-client.el" "/Users/simon/src/CHAT/matrix-client.el/matrix-client.el" nil t)
  require(matrix-client)
  eval((require (quote matrix-client)) nil)
  eval-expression((require (quote matrix-client)) nil nil 127)
  funcall-interactively(eval-expression (require (quote matrix-client)) nil nil 127)
  call-interactively(eval-expression nil nil)
  command-execute(eval-expression)
```

Shouldn't this be a bit easier ?

**Comment by simonmichael** [comment]
(It wants 'notifications, whose provider I'm failing to find.)

**Comment by simonmichael** [comment]
- I try the standalone script, it hangs:
```
~/src/CHAT/matrix-client.el$ ./matrix-client-standalone.el.sh 
2019-11-04 17:22:06.085 emacs[69945:16333651] Failed to initialize color list unarchiver: Error Domain=NSCocoaErrorDomain Code=4864 "*** -[NSKeyedUnarchiver _initForReadingFromData:error:throwLegacyExceptions:]: non-keyed archive cannot be decoded by NSKeyedUnarchiver" UserInfo={NSDebugDescription=*** -[NSKeyedUnarchiver _initForReadingFromData:error:throwLegacyExceptions:]: non-keyed archive cannot be decoded by NSKeyedUnarchiver}
```

- I try quelpa:
```
(require 'quelpa-use-package)
(use-package matrix-client
  :quelpa ((matrix-client :fetcher github :repo "alphapapa/matrix-client.el"
                          :files (:defaults "logo.png" "matrix-client-standalone.el.sh"))))
```
```
Debugger entered--Lisp error: (wrong-type-argument listp \.\.\.)
  #f(compiled-function (elt) #<bytecode 0x4061ba9d>)(\.\.\.)
  mapcar(#f(compiled-function (elt) #<bytecode 0x4061ba9d>) ((emacs "25.1") (anaphora "1.0.0") (dash "2.13.0") (dash-functional "1.2.0") (f "0.17.2") (request "0.2.0") (a "0.1.0") (ov "1.0.6") (s "1.12.0") (tracking "2.9") (esxml "0.3.4") (ht "2.2") \.\.\.))
  package-desc-from-define("matrix-client" "20190728.1931" "A Matrix client" (quote ((emacs "25.1") (anaphora "1.0.0") (dash "2.13.0") (dash-functional "1.2.0") (f "0.17.2") (request "0.2.0") (a "0.1.0") (ov "1.0.6") (s "1.12.0") (tracking "2.9") (esxml "0.3.4") (ht "2.2") \.\.\.)) :keywords (quote ("web" "comm")) :maintainer (quote ("Jay Kamat" . "jaygkamat@gmail.com")) :url "https://github.com/alphapapa/matrix-client.el")
  apply(package-desc-from-define ("matrix-client" "20190728.1931" "A Matrix client" (quote ((emacs "25.1") (anaphora "1.0.0") (dash "2.13.0") (dash-functional "1.2.0") (f "0.17.2") (request "0.2.0") (a "0.1.0") (ov "1.0.6") (s "1.12.0") (tracking "2.9") (esxml "0.3.4") (ht "2.2") \.\.\.)) :keywords (quote ("web" "comm")) :maintainer (quote ("Jay Kamat" . "jaygkamat@gmail.com")) :url "https://github.com/alphapapa/matrix-client.el"))
  package--read-pkg-desc(tar)
  package-tar-file-info()
  quelpa-get-package-desc("/Users/simon/.emacs.d/quelpa/packages/matrix-client-20190728.1931.tar")
  quelpa-package-install((matrix-client :fetcher github :repo "alphapapa/matrix-client.el" :files (:defaults "logo.png" "matrix-client-standalone.el.sh")))
  quelpa((matrix-client :fetcher github :repo "alphapapa/matrix-client.el" :files (:defaults "logo.png" "matrix-client-standalone.el.sh")))
  apply(quelpa (matrix-client :fetcher github :repo "alphapapa/matrix-client.el" :files (:defaults "logo.png" "matrix-client-standalone.el.sh")))
  (if (and quelpa-use-package-inhibit-loading-quelpa (package-installed-p (quote matrix-client))) nil (apply (quote quelpa) (quote ((matrix-client :fetcher github :repo "alphapapa/matrix-client.el" :files (:defaults "logo.png" "matrix-client-standalone.el.sh"))))))
  eval-region(9286 9481 t #f(compiled-function (ignore) #<bytecode 0x4082f35d>))  ; Reading at buffer position 9286
  elisp--eval-defun()
  eval-defun(nil)
  funcall-interactively(eval-defun nil)
  call-interactively(eval-defun nil nil)
  command-execute(eval-defun)
```

**Comment by alphapapa** [comment]
Hi Simon,

I appreciate your interest in this project.  Let me prefix this by saying that this package is intentionally more flexible and/or relaxed in its standards than MELPA packages.  We've considered publishing it on MELPA, and we still might someday, but it would require a lot of work that wouldn't provide any concrete benefits at this time (other than it being on MELPA).  Simply, this package is more fun to work on when I can do things like define macros that aren't prefixed with `matrix-client-`.  And it does need to be fun to work on, because working with Matrix can be tedious.  :)

> I just fixed up my emacs installation and I don't want to risk running some non-standard install script that installs another emacs

Of course, you shouldn't use the "standalone" method if you don't trust it.  But, as the documentation explains, it does not install another Emacs instance.  It uses packages from the `elpa` directory of your default Emacs config.  The code is straightforward, so take a look if you like.  If that's still not your cup of tea, you could use this script to actually run it in a separate Emacs config: https://github.com/alphapapa/alpha-org/blob/master/emacs-sandbox.sh

> I've never heard of quelpa, so I check it out and their README says it's retired - sounds like a dead end

The Quelpa author's retiring from the project is news to me.  Since I use and recommend Quelpa, I suspect this means I'll have to adopt the package myself.  *sigh*  In the meantime, it works fine, so feel free to use it, especially with `quelpa-use-package`.

> So it's manual installation for me. I try my usual fumble-around-with-dired technique: open the directory, select all .el files, B, L. 6 files fail to compile. The log shows lots of warnings and errors but I can't see which packages it might be needing.

> I read and follow the manual instructions more carefully. I add the matrix-client.el directory to load-path and try (require 'matrix-client). Now I see it's failing on require(ov). I work through these errors / look at the source and install a bunch of things with list-packages: a, anaphora, dash-functional, esxml, ov, request tracking.

That's because you tried to compile the Elisp files without having the dependencies installed, which Quelpa installs automatically.

>     ~/src/CHAT/matrix-client.el$ ./matrix-client-standalone.el.sh 
>     2019-11-04 17:22:06.085 emacs[69945:16333651] Failed to initialize color list unarchiver: Error Domain=NSCocoaErrorDomain Code=4864 "*** -[NSKeyedUnarchiver _initForReadingFromData:error:throwLegacyExceptions:]: non-keyed archive cannot be decoded by NSKeyedUnarchiver" UserInfo={NSDebugDescription=*** -[NSKeyedUnarchiver _initForReadingFromData:error:throwLegacyExceptions:]: non-keyed archive cannot be decoded by NSKeyedUnarchiver}

I have no idea what that is about.  Something related to MacOS, I guess?  How and from where did you install your Emacs?

> I try quelpa:

>     Debugger entered--Lisp error: (wrong-type-argument listp \.\.\.) ...

*shrug*  I don't know, the version of Quelpa I have installed works fine.  I recommend trying `emacs -q` to ensure it's not a problem with your config.

> (It wants 'notifications, whose provider I'm failing to find.)

My Emacs shows:

    notifications is a built-in package.

         Status: Built-In.
        Summary: Client interface to desktop notifications.


    This package provides an implementation of the Desktop Notifications
    <http://developer.gnome.org/notification-spec/>.

    In order to activate this package, you must add the following code
    into your .emacs:

      (require 'notifications)

    For proper usage, Emacs must be started in an environment with an
    active D-Bus session bus.

In conclusion?  Well...

1.  This package has several dependencies which must be installed first.
2.  Since it's not published in an ELPA repo, Quelpa serves as the alternative.
3.  AFAIK Quelpa still works properly, despite its author retiring a few weeks ago.
4.  If that's not the case, we'll need to find another alternative (or I'll have to fix Quelpa myself).  You could probably use `straight.el` instead of Quelpa, but I don't use it myself, so I can't offer help with it.  `feather` is another alternative package manager you might like.
5.  As a fallback, you could `grep Package-Requires *.el` and install the dependencies manually, after which byte-compilation and loading should work (there may be warnings, but they can be ignored).

Please let me know how it goes.  Thanks.

**Comment by simonmichael** [comment]
Thanks for the help @alphapapa, especially the clue about notifications being built in. Wouldn't you know it, after restarting emacs this is working:
```
(setq load-path (cons (expand-file-name "~/src/CHAT/matrix-client.el") load-path))
(require 'matrix-client)
```
The rest above is just feedback on what installation can be like for some users, I hope it's helpful. I'll move on to next issue ("Syncing" for a long time then fails, I expect I'll figure it out).

**Comment by lebensterben** [comment]
@alphapapa Just FYI, I'm using spacemacs and I modified the lines for installation with `quelpa-use-package` to the following:

```elisp
(matrix-client :location (recipe :fetcher github
								 :repo "alphapapa/matrix-client.el"
								 :files (:defaults "logo.png" "matrix-client-standalone.el.sh")))
```

And put this in `dotspacemacs-additional-packages '(XXXXX)`.

It turns out that it can work perfectly fine.

**Comment by alphapapa** [comment]
Closing issues and archiving repository.  Please see https://github.com/alphapapa/ement.el for the new client.
